src/zasm.zig

```diff
@@ -0,0 +1,716 @@
+// MIT License
+//
+// Copyright (c) 2020-2022 Lee Cannon
+
+const std = @import("std");
+
+// ~~~ [ zig-x86_64/src/registers/control.zig ] ~~~
+
+/// Various control flags modifying the basic operation of the CPU.
+pub const Cr0 = packed struct {
+    /// Enables protected mode.
+    protected_mode: bool,
+
+    /// Enables monitoring of the coprocessor, typical for x87 instructions.
+    ///
+    /// Controls together with the `task_switched` flag whether a `wait` or `fwait`
+    /// instruction should cause an `#NE` exception.
+    monitor_coprocessor: bool,
+
+    /// Force all x87 and MMX instructions to cause an `#NE` exception.
+    emulate_coprocessor: bool,
+
+    /// Automatically set to 1 on _hardware_ task switch.
+    ///
+    /// This flags allows lazily saving x87/MMX/SSE instructions on hardware context switches.
+    task_switched: bool,
+
+    /// Indicates support of 387DX math coprocessor instructions.
+    ///
+    /// Always set on all recent x86 processors, cannot be cleared.
+    extension_type: bool,
+
+    /// Enables the native (internal) error reporting mechanism for x87 FPU errors.
+    numeric_error: bool,
+
+    z_reserved6_15: u10,
+
+    /// Controls whether supervisor-level writes to read-only pages are inhibited.
+    ///
+    /// When set, it is not possible to write to read-only pages from ring 0.
+    write_protect: bool,
+
+    z_reserved17: bool,
+
+    /// Enables automatic usermode alignment checking if `RFlags.alignment_check` is also set.
+    alignment_mask: bool,
+
+    z_reserved19_28: u10,
+
+    /// Ignored. Used to control write-back/write-through cache strategy on older CPUs.
+    not_write_through: bool,
+
+    /// Disables some processor caches, specifics are model-dependent.
+    cache_disable: bool,
+
+    /// Enables paging.
+    ///
+    /// If this bit is set, `protected_mode` must be set.
+    paging: bool,
+
+    z_reserved32_63: u32,
+
+    /// Read the current set of CR0 flags.
+    pub fn read() Cr0 {
+        return Cr0.fromU64(readRaw());
+    }
+
+    /// Read the current raw CR0 value.
+    fn readRaw() u64 {
+        return asm ("mov %%cr0, %[ret]"
+            : [ret] "=r" (-> u64),
+        );
+    }
+
+    /// Write CR0 flags.
+    ///
+    /// Preserves the value of reserved fields.
+    pub fn write(self: Cr0) void {
+        writeRaw(self.toU64() | (readRaw() & ALL_RESERVED));
+    }
+
+    /// Write raw CR0 flags.
+    ///
+    /// Does _not_ preserve any values, including reserved fields.
+    fn writeRaw(value: u64) void {
+        asm volatile ("mov %[val], %%cr0"
+            :
+            : [val] "r" (value),
+            : "memory"
+        );
+    }
+
+    const ALL_RESERVED: u64 = blk: {
+        var flags = std.mem.zeroes(Cr0);
```

Comment:
`std.mem.zeroes` is equivalent to `memset`, but instead of a size, it takes a type and figures out how to represent a zero initialized value of that type.